CovertInfiltration/Src/CovertInfiltration/Classes/XComGameState_SquadPickupPoint.uc

```diff
@@ -112,37 +113,45 @@ simulated function ConfirmExfiltrate()
 simulated function PayThePrice(XComGameState NewGameState)
 {
 	local XComGameState_Item Intel;
+	local ArtifactCost Cost;
 
 	Intel = `XCOMHQ.GetItemByName('Intel');
 
 	Intel = XComGameState_Item(NewGameState.ModifyStateObject(class'XComGameState_Item', Intel.GetReference().ObjectID));
-	Intel.Quantity -= IntelCost;
+	
+	foreach ExfiltrateCost.ResourceCosts(Cost)
+	{
+		if (Cost.ItemTemplateName == 'Intel')
+		{
+			Intel.Quantity -= Cost.Quantity;
+			break;
+		}
+	}
 }
 
 simulated function ClearUnitsFromAction(XComGameState NewGameState)
 {
 	local XComGameState_CovertAction CovertAction;
-	local XComGameState_StaffSlot StaffSlot;
+	local XComGameState_StaffSlot SlotState;
 	local XComGameState_Unit Unit;
 	local XComGameStateHistory History;
-	local StateObjectReference UnitRef;
-	local array<StateObjectReference> Squad;
 	local CovertActionStaffSlot CovertActionSlot;
 
 	History = `XCOMHISTORY;
 	CovertAction = XComGameState_CovertAction(History.GetGameStateForObjectID(ActionRef.ObjectID));
-	Squad = class'X2Helper_Infiltration'.static.GetCovertActionSquad(CovertAction);
-
-	foreach Squad(UnitRef)
-	{
-		Unit = XComGameState_Unit(NewGameState.ModifyStateObject(class'XComGameState_Unit', UnitRef.ObjectID));
-		class'X2Helper_Infiltration'.static.CreateWillRecoveryProject(NewGameState, Unit);
-	}
 
 	foreach CovertAction.StaffSlots(CovertActionSlot)
 	{
-		StaffSlot = XComGameState_StaffSlot(History.GetGameStateForObjectID(CovertActionSlot.StaffSlotRef.ObjectID));
-		StaffSlot.EmptySlot(); // This is noop if the slot is empty
+		SlotState = XComGameState_StaffSlot(History.GetGameStateForObjectID(CovertActionSlot.StaffSlotRef.ObjectID));
+		if (SlotState.IsSlotFilled())
+		{
+			Unit = SlotState.GetAssignedStaff();
+			if (Unit.IsSoldier() && Unit.UsesWillSystem())
+			{
+				class'X2Helper_Infiltration'.static.CreateWillRecoveryProject(NewGameState, Unit);
+			}
+			SlotState.EmptySlot();
```

Comment:
Pass `NewGameState` here as otherwise you should get redscreens about out-of-order submissions